unison-cli/src/Unison/CommandLine/Main.hs

```diff
@@ -98,10 +99,32 @@ main ::
   Codebase IO Symbol Ann ->
   Maybe Server.BaseUrl ->
   UCMVersion ->
-  ((Branch IO, Path.Absolute) -> IO ()) ->
+  (Branch IO -> STM ()) ->
+  (Path.Absolute -> STM ()) ->
   IO ()
-main dir welcome initialPath (config, cancelConfig) initialInputs runtime sbRuntime codebase serverBaseUrl ucmVersion notifyChange = do
-  root <- Codebase.getRootBranch codebase
+main dir welcome initialPath (config, cancelConfig) initialInputs runtime sbRuntime codebase serverBaseUrl ucmVersion notifyBranchChange notifyPathChange = Ki.scoped \scope -> do
+  rootVar <- newEmptyTMVarIO
+  initialRootCausalHash <- Codebase.getRootCausalHash codebase
+  _ <- Ki.fork scope $ do
+    root <- Codebase.getRootBranch codebase
+    atomically $ do
+      -- Try putting the root, but if someone else as already written over the root, don't
+      -- overwrite it.
+      void $ tryPutTMVar rootVar root
+    -- Start forcing the thunk in a background thread.
+    -- This might be overly aggressive, maybe we should just evaluate the top level but avoid
+    -- recursive "deep*" things.
+    void $ UnliftIO.evaluate root
+  let initialState = Cli.loopState0 initialRootCausalHash rootVar initialPath
+  Ki.fork_ scope $ do
+    let loop lastRoot = do
```

Comment:
Is it okay that this notify might miss updates? If so I think it could use a comment somewhere.